### ISSUE TITLE
Unhandled Exception in HardwareUtils.GenerateMachineID()

### ISSUE THREAD

**Issue by felipe19930** [SteamRE/SteamKit#629]
So I was trying to find a way to run [ArchiSteamFarm](https://github.com/JustArchiNET/ArchiSteamFarm) on my shitty Android device and fortunately I was able to by using Termux + Debian image, via chroot. I installed dotnet-core, configured ASF, yada yada. When I ran ASF I got this:

<details>
 <summary>log.txt</summary>

```
2019-01-22 02:19:51|dotnet-18013|INFO|ASF|InitASF() ArchiSteamFarm V4.0.0.0 (generic/a9987dee-cb0c-4f02-a592-07e5dfeeeba3 | Linux 3.4.113-g22bc4ed7f41 #1 SMP PREEMPT Wed Jan 16 03:40:14 UTC 2019)
2019-01-22 02:19:56|dotnet-18013|INFO|ASF|Start() Starting IPC server...
2019-01-22 02:20:05|dotnet-18013|INFO|ASF|Start() IPC server ready!
2019-01-22 02:20:05|dotnet-18013|INFO|ASF|InitPlugins() Initializing Plugins...
2019-01-22 02:20:05|dotnet-18013|INFO|ASF|InitPlugins() Nothing found!
2019-01-22 02:20:05|dotnet-18013|WARN|ASF|InitBots() You're running more personal bot accounts than our upper recommended limit (10). Be advised that this setup is not supported and might cause various Steam-related issues, including account suspensions. Check out the FAQ for more details.
2019-01-22 02:22:35|dotnet-18013|INFO|1|Start() Starting...
2019-01-22 02:22:35|dotnet-18013|INFO|1|Connect() Connecting...
2019-01-22 02:22:37|dotnet-18013|INFO|1|OnConnected() Connected to Steam!
2019-01-22 02:22:37|dotnet-18013|INFO|1|OnConnected() Logging in...
2019-01-22 02:22:40|dotnet-18013|FATAL|ASF|OnUnhandledException() System.AggregateException: One or more errors occurred. (An error was encountered while querying information from the operating system.) ---> System.Net.NetworkInformation.NetworkInformationException: An error was encountered while querying information from the operating system.
   at System.Net.NetworkInformation.LinuxNetworkInterface.GetLinuxNetworkInterfaces()
   at SteamKit2.DefaultInfoProvider.GetMacAddress()
   at SteamKit2.HardwareUtils.GenerateMachineID()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__278_1(Object obj)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait(TimeSpan timeout)
   at SteamKit2.HardwareUtils.GetMachineID()
   at SteamKit2.SteamUser.LogOn(LogOnDetails details)
   at ArchiSteamFarm.Bot.OnConnected(ConnectedCallback callback)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.<>c.<ThrowAsync>b__7_1(Object state)
   at System.Threading.QueueUserWorkItemCallback.<>c.<.cctor>b__6_0(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.QueueUserWorkItemCallback.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
---> (Inner Exception #0) System.Net.NetworkInformation.NetworkInformationException (0x80004005): An error was encountered while querying information from the operating system.
   at System.Net.NetworkInformation.LinuxNetworkInterface.GetLinuxNetworkInterfaces()
   at SteamKit2.DefaultInfoProvider.GetMacAddress()
   at SteamKit2.HardwareUtils.GenerateMachineID()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__278_1(Object obj)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)<---

2019-01-22 02:22:40|dotnet-18013|ERROR|ASF|Exit() Exiting with nonzero error code!
```

</details>
<br>

I don't know what causes this (I'm not a C# person, **at all**), but I was able to fix this by simply catching `NetworkInformationException`. I'll send a PR but feel free to disregard it if you have a better fix, of course.

Thank you!

**Comment by JustArchi** [comment]
This indeed looks like SK2 issue, so I'm not mad that you didn't report to me first 😃. It's probably connected with the fact that linux kernel used in Android is very specific (I'm shocked that dotnet-core even works in debian chroot to be honest), and SK2 code probably never considered the possibility of network interfaces being unavailable. You didn't specify if you ran ASF with root - normal users are so limited within android ecosystem that I'm not shocked at all that SK2 (corefx to be precise) throws unhandled exception.

Glad to see a PR already opened for it, good job 👍

**Comment by felipe19930** [comment]
Thanks!

ASF works perfectly well, judging by the 30 minutes I kept it running (about 40 bots, plus a very slow ARM processor, plus 1 GB of RAM). Explaining a bit better how I did it:

I've used [AnLinux](https://github.com/EXALAB/AnLinux-App) to set-up a Debian environment using [Termux](https://github.com/termux/termux-app) (everything as a regular user, no root access). After that, I simply chroot'd into Debian and downloaded .NET core for ARM32. Now, whilst on Debian, I was running everything as root, but I assume that doesn't matter much since it can only see/modify things within the Debian filesystem and it can't touch Android's (?)

Bonus image of ASF and ASF-ui running and working locally (I can access it remotely too, except for the commands page, which I haven't looked into yet to be honest):

![screenshot_20190122-023527](https://user-images.githubusercontent.com/35700659/51512783-ddac7680-1dee-11e9-8ebb-506323235a71.png)
